File: basic-algorithms/10chunkymonkey.go
// https://play.golang.org/p/HrFh0tQC-M
package main

import (
	"fmt"
)

func chunkArrayInGroups(arr []rune, size int) (bnoden [][]rune) {

	temp := []rune{}

	for i := 0; i < len(arr); i++ {
		temp = append(temp, arr[i])
		if i%size == size-1 {
			bnoden = append(bnoden, temp[0:len(temp)])
			temp = []rune{}
		}
	}
	if len(temp) != 0 {
		bnoden = append(bnoden, temp[0:len(temp)])
	}
	return bnoden
}

// Test
func main() {
	arr1 := []rune{'a', 'b', 'c', 'd'}
	arr2 := []rune{0, 1, 2, 3, 4, 5}
	arr3 := []rune{0, 1, 2, 3, 4, 5}
	arr4 := []rune{0, 1, 2, 3, 4, 5}
	arr5 := []rune{0, 1, 2, 3, 4, 5, 6}
	arr6 := []rune{0, 1, 2, 3, 4, 5, 6, 7, 8}
	arr7 := []rune{0, 1, 2, 3, 4, 5, 6, 7, 8}

	fmt.Printf("%c\n", chunkArrayInGroups(arr1, 2)) // [[a b] [c d]]
	fmt.Println(chunkArrayInGroups(arr2, 3))        // [[0 1 2] [3 4 5]]
	fmt.Println(chunkArrayInGroups(arr3, 2))        // [[0 1] [2 3] [4 5]]
	fmt.Println(chunkArrayInGroups(arr4, 4))        // [[0 1 2 3] [4 5]]
	fmt.Println(chunkArrayInGroups(arr5, 3))        // [[0 1 2] [3 4 5] [6]]
	fmt.Println(chunkArrayInGroups(arr6, 4))        // [[0 1 2 3] [4 5 6 7] [8]]
	fmt.Println(chunkArrayInGroups(arr7, 2))        // [[0 1] [2 3] [4 5] [6 7] [8]]
}

File: basic-algorithms/js/10-chunky-monkey.js
function chunkArrayInGroups(arr, size, bnoden, temp) {
	bnoden = [], temp = []
	for (var i in arr) {
		temp.push(arr[i])
		if (i % size === size-1)
			bnoden.push(temp), temp = []
	}
	if (temp.length !== 0) bnoden.push(temp)

	// Only for testing
	function test() {
		for (var i in bnoden)
			console.log(bnoden[i])
	}
	return test(bnoden) // Only for testing

	// return bnoden
}

// Test
console.log(chunkArrayInGroups(["a", "b", "c", "d"], 2)) // [["a", "b"], ["c", "d"]]
console.log(chunkArrayInGroups([0, 1, 2, 3, 4, 5], 3)) // [[0, 1, 2], [3, 4, 5]]
console.log(chunkArrayInGroups([0, 1, 2, 3, 4, 5], 2)) // [[0, 1], [2, 3], [4, 5]]
console.log(chunkArrayInGroups([0, 1, 2, 3, 4, 5], 4)) // [[0, 1, 2, 3], [4, 5]]
console.log(chunkArrayInGroups([0, 1, 2, 3, 4, 5, 6], 3)) // [[0, 1, 2], [3, 4, 5], [6]]
console.log(chunkArrayInGroups([0, 1, 2, 3, 4, 5, 6, 7, 8], 4)) // [[0, 1, 2, 3], [4, 5, 6, 7], [8]]
console.log(chunkArrayInGroups([0, 1, 2, 3, 4, 5, 6, 7, 8], 2)) // [[0, 1], [2, 3], [4, 5], [6, 7], [8]]